### PR TITLE
Fix to #32972 - The database default created by Migrations for primitive collections mapped to JSON is invalid

### LINKED ISSUE #32972: The database default created by Migrations for primitive collections mapped to JSON is invalid
See https://github.com/dotnet/efcore/issues/32896#issuecomment-1911969614

### PR DESCRIPTION
**Pull request by maumar**
Problem was that when adding new required column to an existing table we always need to provide default value (to fill the values for rows already in the table). For collection of primitives that get map to json we should be setting the value to empty JSON collection, i.e. '[]' but we were not doing that.

Fixes #32972

### CODE REVIEW

File: src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs [hunk 1]
```diff
@@ -1243,7 +1243,13 @@ protected virtual IEnumerable<MigrationOperation> Remove(IColumn source, DiffCon
...
         if (!column.TryGetDefaultValue(out var defaultValue))
         {
-            defaultValue = null;
+            // for non-nullable collections of primitives that are mapped to JSON we set a default value corresponding to empty JSON collection
+            defaultValue = !inline
+                && column is { IsNullable: false, StoreTypeMapping: { ElementTypeMapping: not null, Converter: ValueConverter columnValueConverter } }
+                && columnValueConverter.GetType() is Type { IsGenericType: true } columnValueConverterType
+                && columnValueConverterType.GetGenericTypeDefinition() == typeof(CollectionToJsonStringConverter<>)
+                ? "[]"
+                : null;
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs:1252]: Move the `defaultValue` logic below here as well, so it's in the same place

**Review by AndriySvyryd — APPROVED** [review]
So we do this already for JSON columns mapped to owned entity types?

**Comment by maumar** [comment]
yes, we already do this for JSON InitializeJsonColumn. Only need to do this for references - JSON collections can never be required

File: src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs [hunk 2]
```diff
@@ -1243,9 +1243,20 @@ protected virtual IEnumerable<MigrationOperation> Remove(IColumn source, DiffCon
 
         if (!column.TryGetDefaultValue(out var defaultValue))
         {
-            defaultValue = null;
+            // for non-nullable collections of primitives that are mapped to JSON we set a default value corresponding to empty JSON collection
+            defaultValue = !inline
```
Comment (roji) [inline src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs:1247]: Should this logic move into GetDefaultValue instead, which is the code responsible for determining default values? It would have to accept the column as opposed to just the CLR type, of course.

**Review by roji — COMMENTED** [review]
I know this implementation is in line with how things are currently done, but I'm wondering whether implementing the default value in the differ (as opposed to the migrations SQL generator) as the right place... Doesn't this mean that we're setting the default value even when simply a table is created with a JSON collection (as opposed to a collection column being added to an existing table)? In that case having a default value doesn't seem appropriate...

In other words, should this all happen in the AddColumn case in SqlMigrationsGenerator instead?

File: src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs:1247]: This code looks messy because it's conflating the default value that should be used when a new row is added with the default value that should be used when a new required column is added or an existing column is made required.

Comment (ajcvickers) [reply src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs:1247]: @roji This is a long standing issue with the way Migrations uses default values. See #27084.
